src/providers/duplicatelabels.ts

```diff
@@ -0,0 +1,101 @@
+import * as vscode from 'vscode'
+import * as path from 'path'
+
+import type {Extension} from '../main'
+
+type LabelEntries = {
+    [key: string]: {
+        file: string,
+        range: vscode.Range
+    }[]
+}
+
+export class DuplicateLabels {
+    private readonly extension: Extension
+    private readonly duplicatedLabelsDiagnostics = vscode.languages.createDiagnosticCollection('Duplicate Labels')
+    private labelEntries: LabelEntries = {}
```

Comment:
Why is `labelEntries` a property of `DuplicateLabels` class? How about making `parseFileLabels` return it and `showDiagnostics` called with it as an argument?

Reply:
`labelEntries` is a property of `DuplicateLabels` because it holds the labels for all the project files and we only update the labels defined in the file from which the check was triggered.

> I don't think caching them here has benefits. It is enough to use `this.extension.manager.cachedContent[file].element.reference` every time.

I do not really understand your point. Can you expand a little?

Reply:
We should avoid using caches if possible. Using caches requires us to ensure caches to be consistent with updates. We easily fail to achieve that, which leads to bugs.

I think it can be acceptable even if the code becomes much slower.

Reply:
I do feel concerned by the computational cost of every feature, which led me to adopt a partial update approach. However, to deal with the removal of inclusions as `\include{myfile}`, we need to remove any references to files not in `manager.getIncludedTeX()`

https://github.com/James-Yu/LaTeX-Workshop/blob/292fe7ee0258441f55a7bf7f9e25fbf0096f24e0/src/providers/duplicatelabels.ts#L55-L58

Then, I am not sure it still makes sense to adopt a per file approach rather than recomputing `labelEntries` every time.

If I understood you well, this is also your point, isn't it?

Reply:
I am saying that we don't have to feel concerned by the computational cost here. We should be concerned about whether the changes lead to bugs here.

At least, we should stop adding cache here and there ad hoc. If you want to add `labelEntries` somewhere, it seems to me appropriate that `labelEntries` becomes a property of `this.extension.manager`. We should update `labelEntries` every time `this.extension.manager.cachedContent[file].element.reference` updated.

Reply:
Another solution is that extracting a class having the `labelEntries` property from `DuplicateLabels` class into `src/components/managerlib/`. 

I guess that we will have to access `labelEntries` for purposes other than duplicate labels.

Reply:
I have simplified the computation of duplicates and do not use cache anymore.